# Exploring eBay Car Sales Data

## Introduction

In this project, we will work with a dataset of used cars from *eBay Kleinanzeigen*, a classifieds section of the German eBay website. The original dataset was scraped and uploaded to Kaggle; it is not available there anymore but can still be found [here](https://data.world/data-society/used-cars-data). It is important to note that from the full dataset, we have sampled 50,000 data points to facilitate faster code running.

The aim of this project is to clean the data and analyze the included used car listings. We'll also become familiar with some of the unique benefits Jupyter Notebook provides for pandas.

## Getting Started

First we will import the pandas and NumPy libraries:

In [1]:
import pandas as pd
import numpy as np

And then we will read the `autos.csv` file into pandas.

In [2]:
autos = pd.read_csv('autos.csv', encoding="Latin-1")

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

When I first tried to read `autos.csv` into pandas, I encountered an encoding error because I did not specify any encoding (i.e. I left that parameter blank). Because I did this, pandas defaulted to using `UTF-8` but failed since the file was not encoded with that format. Because I didn't know the file's encoding, I specified with another common format, `Latin-1`, and luckily I was then able to read the file without error.

The above block highlights a neat feature of Jupyter Notebook. I simply populated a cell with the variable `autos`, a DataFrame, and Jupyter then rendered the first and last five rows upon running the cell. It does this with any pandas object. Now we're going to seek out some more information about `autos`.

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Running `autos.info()` provides some useful at-a-glance information.
- Some columns contain null values, specifically `vehicleType`, `gearbox`, `model`, `fuelType`, and `notRepairedDamage`. However, the amount never exceeds 20%.
- Most columns contain information of the object (string) type, but some have numerical values.
    - `yearOfRegistration`, `powerPS`, `monthOfRegistration`, `nrOfPictures`, and `postalCode` are of the int64 type.
    - The dtypes line at the bottom do indeed confirm there are five columns of the int64 type and no float64 types.
- There are 20 columns in this DataFrame. The column names use [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case), which means we can't just replace spaces with underscores.

I have printed the first five rows again for convenience using `autos.head()` and here we find more information that will further round out our understandings of the data.
- As the original compiler noted, the text is in German, so we will have to translate it all for an English-speaking audience.
- The `price` and `odometer` columns contain numerical values stored as strings, which we will target when we clean the data. Some date columns are stored as strings as well.

## Cleaning Column Names

Let's start by converting the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive. For reference, here is the data dictionary:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing.
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month of the year in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test', 'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code', 'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In the cells above, I first used `autos.columns` to print an array of the existing columns. To make the edits, I created a list of column names that converts from camelcase to snakecase wherever applicable and reworded some names for clarity's sake. I then assigned the modified column names back to `autos.columns`. Finally, I printed the first five rows again to confirm the necessary changes were successfully implemented.

## Initial Exploration and Cleaning

Now we're going to do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:
- As mentioned earlier, examples of numeric data stored as text.
- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.

Let's take a look at descriptive statistics for all columns. The `DataFrame.describe()` method will do this for us, and the argument `include='all'` will get both categorical and numerical columns.

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Some findings:
- Looking at the `unique` and `freq` rows, we see that the columns `seller` and `offer_type` consist of mostly one value and are candidates to be dropped.
- The `num_photos` column has `NaN` in the `unique` row where we would expect to see a numerical value. This merits further investigation.
- Also as mentioned earlier, the `price` and `odometer` columns are stored as text.

We'll start off by taking a closer look at the `num_photos` column.

In [8]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

This column looks to be filled with `0` on every row. We'll drop this one too.

In [9]:
autos = autos.drop(["seller", "offer_type", "num_photos"], axis=1)

Now we'll clean and convert `price` and `odometer`.

In [10]:
autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [11]:
autos["odometer"] = (autos["odometer"]
                             .str.replace("km","")
                             .str.replace(",","")
                             .astype(int)
                             )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

## Exploring the Odometer and Price Columns

We're going to continue exploring the data in these columns, looking in particular for anything that may not look right. Generally, it's a good idea to use the following to get started on data exploration in pandas:

- `Series.unique().shape` to see how many unique values
- `Series.describe()` to view min/max/mean etc.
- `Series.value_counts()`, with some variations:
    - chained to `.head()` if there are lots of values
    - Because `Series.value_counts()` returns a series, we can use `Series.sort_index()` with `ascending=` `True` or `False` to view the highest and lowest values with their counts (can also chain to `.head()` here)
- When removing outliers, we can do `df[(df["col"] >= x ) & (df["col"] <= y )]`, but it's more readable to use `df[df["col"].between(x,y)]`

In [12]:
# Start with the `odometer_km` column

autos["odometer_km"].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

The values in this column are rounded, which indicates that sellers may have had to choose from preset options. There are also far more high mileage than low mileage vehicles. This dataset appears to be fairly simple so we'll move onto the next column.

In [13]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
autos["price"].value_counts().head(20).sort_index(ascending=False)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


4500     394
3500     498
2500     643
2200     382
2000     460
1500     734
1200     639
1000     639
999      434
950      379
900      420
850      410
800      498
750      433
700      395
650      419
600      531
500      781
300      384
0       1421
Name: price, dtype: int64

Because this column consists of 2357 unique values, I chained `head()` to `value_counts()`, passing 20 rows as the argument to still get a large enough picture of the data.

Looking at the descriptive statistics, we see that this column has a mean of \\$984 for the price of a used car. The minimum is \\$0, which seems odd. In fact, our value counts indicate there are 1421 entries with \\$0 for the price. This is less than 3\% of all entries, so we'll probably omit those rows for analysis.

However, the maximum price is one hundred million dollars. This is even more odd. We'll shift our attention there now, and look for any other possible outliers.

In [14]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [15]:
# Look at lowest prices too

autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

Now we must determine which values are outliers to omit and which values to keep for meaningful analysis.

As stated earlier, anything with a price of \\$0 will go. There are lots of other values on the low end, but given that eBay is an auction site, it is entirely possible some sellers will put their starting bids that low. So, we will keep even those with a price of \\$1.

For the high end, the prices increase somewhat steadily until \\$350,000. After that the next highest value is \\$999,990. Almost all the prices above this seem arbitrary and nonsensical, so we'll decide to omit everything above \\$350,000.

Let's take a brief look again at the remaining data.

In [16]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## Exploring the Date Columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website
    
Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into numerical representations so we can understand them quantitatively.

In [17]:
autos[["date_crawled", "last_seen", "ad_created"]][0:5]

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

In all three columns, the first 10 characters represent the date. This should be sufficient if we are trying to understand date ranges. Now we're going to look at the distribution of values in each string column.

In [18]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [19]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values()
        )

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

This calculations displays percentages instead of counts. The date with the most crawls, at almost 4% of the total, is April 3, 2016, represented by the date stamp `2016-04-03`. No date takes a lion's share of the crawling, with most days between 1-4%, and the crawling was all done in March and early April 2016.

In [20]:
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

In [21]:
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values()
        )

2016-01-07    0.000021
2016-01-03    0.000021
2016-02-09    0.000021
2016-01-29    0.000021
2015-12-05    0.000021
                ...   
2016-03-12    0.036755
2016-04-04    0.036858
2016-03-21    0.037579
2016-03-20    0.037949
2016-04-03    0.038855
Name: ad_created, Length: 76, dtype: float64

The oldest ads (as of the crawling period) are from June 11, 2015 and go through until April 7, 2016. There are also very few ads more than several months old, possibly indicating it usually takes at least a month or two to make a sale. The top 5 days for amount of ads posted are all in March and early April 2016, with each day claiming almost 4% of the total amount of ads in the dataset.

In [22]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [23]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values()
        )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-18    0.007351
2016-03-08    0.007413
2016-03-13    0.008895
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-14    0.012602
2016-03-27    0.015649
2016-03-19    0.015834
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-26    0.016802
2016-03-23    0.018532
2016-03-25    0.019211
2016-03-24    0.019767
2016-03-21    0.020632
2016-03-20    0.020653
2016-03-28    0.020859
2016-03-22    0.021373
2016-03-29    0.022341
2016-04-01    0.022794
2016-03-12    0.023783
2016-03-31    0.023783
2016-04-04    0.024483
2016-03-30    0.024771
2016-04-02    0.024915
2016-04-03    0.025203
2016-03-17    0.028086
2016-04-05    0.124761
2016-04-07    0.131947
2016-04-06    0.221806
Name: last_seen, dtype: float64

The `last_seen` column records when the crawler last saw a posted ad. The date range for this is similar to that of the `date_crawled` column, through March and early April 2016. However, the crawler last saw most ads toward the end of its crawling period, with 22% of ads last seen on its penultimate day, April 6, and another 13% on its last day. It last saw very few ads toward the beginning of the crawl period.

`registration_year` already has numerical data, so we're now going to use `Series.describe()` to understand the distribution of this column.

In [24]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The minimum value (earliest year) is 1000. Cars did not exist then, so it has to be an error, which then affects the mean and standard deviation calculations. The maximum value (latest year) is 9999, another error.

## Dealing with Incorrect Registration Year Data

Now it's time to decide the highest and lowest applicable values for the `registration_year` column.

Because a car can't be first registered after its listing was seen, any vehicle with a registration year after 2016 is definitely inaccurate. As for the earliest valid year, it could realistically go as far back as the early 1900s. To cast a wide enough yet reasonable net, we'll count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely.

In [25]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.038793369710697

This is less than 4% of the data, so we will remove these rows.

In [26]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

Based on the cleaned data, most cars had been first registered in the last 20 years before the ad posting, with most cars being between 10-20 years old.

## Exploring Price by Brand

This part will focus on aggregation - that is, where we apply a statistical operation to groups of our data. In our case, we will be exploring variations across different car brands, so our attention will now turn to the `brand` column. Specifically, we want to find the mean selling price for certain brands.

In [27]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

The top five brands are all German manufacturers, and the top four alone make up more than 50% of all listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [28]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [29]:
brand_mean_prices = {}

for brand in common_brands:
    selected_brand = autos[autos["brand"] == brand]
    mean = selected_brand["price"].mean()
    brand_mean_prices[brand] = int(mean)
    
brand_mean_prices

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

To recap this process, we:

1. Used the `Series.index` attribute to access a list of only the brands we've chosen to aggregate on
2. Assigned the resulting list to a variable `common_brands`
3. Initialized an empty dictionary `brand_mean_prices` to hold the aggregate data
4. Looped over `common_brands`
    1. Selected only the rows with a brand name equal to the current iteration, using boolean comparison
    2. Calculated the mean selling price for just those rows using information from the `price` column
    3. Assigned the mean value to the dictionary, using the brand name as the key
    
As for the aggregate data, we can make out three categories of mean prices. In the high category are brands BMW, Mercedes-Benz, and Audi; the low category has Opel and Ford; Volkswagen finds itself in the middle. Perhaps that is why Volkswagen makes up such a large share of ads - it may offer a happy medium of both price and quality.

## Storing Aggregate Data in a DataFrame

Our next step is to use aggregation again to understand the average mileage for cars in the top 6 brands and see if there is any visible link with mean price.

While we could always visually compare both aggregated series objects, we will instead combine the series objects into a single dataframe with a shared index. This method will allow us to make more precise and accurate observations. To accomplish this, we will be learning two pandas methods: the series constructor and the dataframe constructor.

Before we get to that, we'll first calculate the mean mileage for each of the top brands and store the results in a dictionary. (Yes, I know it's actually recorded in kilometers here.)

In [30]:
brand_mileage = {}

for brand in common_brands:
    selected_brand = autos[autos["brand"] == brand]
    mean_mileage = selected_brand["odometer_km"].mean()
    brand_mileage[brand] = int(mean_mileage)
    
brand_mileage

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

Now it's time to convert both dictionaries into series objects using the series constructor.

In [31]:
bmp_series = pd.Series(brand_mean_prices)
bmp_series

volkswagen       5402
bmw              8332
opel             2975
mercedes_benz    8628
audi             9336
ford             3749
dtype: int64

In [32]:
bmm_series = pd.Series(brand_mileage)
bmm_series

volkswagen       128707
bmw              132572
opel             129310
mercedes_benz    130788
audi             129157
ford             124266
dtype: int64

The keys in the dictionary became the index in the series object. Now we're going to create a single-column dataframe from this series object. We need to use the `columns` parameter when calling the dataframe constructor to specify the column name, or else the column name will be set to `0` by default.

In [33]:
brand_df = pd.DataFrame(bmp_series, columns=["mean price"])
brand_df

mean price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749

Excellent, now all we have to do is assign the other series as a new column in this dataframe.

In [34]:
brand_df["mean mileage"] = bmm_series
brand_df

mean price  mean mileage
volkswagen           5402        128707
bmw                  8332        132572
opel                 2975        129310
mercedes_benz        8628        130788
audi                 9336        129157
ford                 3749        124266

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.

## Conclusion

This project was an exercise in exploring Jupyter Notebook's pandas functionalities. We practiced applying a variety of pandas methods to explore and understand a data set on car listings.

Of course, there are many more steps we could take to uncover even more information about the data. Some suggestions will follow, and we will return to this project to complete them as we continue to build our capabilities in pandas. For example, there were observations we made about the data during the cleaning phase that we did not follow up on throughout the rest of the sections.

- Data cleaning next steps:
    - Identify categorical (string) data that uses German words, translate them, and map their values to their English counterparts
    - Convert the dates to uniform numeric data, so `"2016-03-21"` becomes the integer `20160321`
    - See if there are particular keywords in the name column that we can extract as new columns
- Analysis next steps:
    - Find the most common brand/model combinations
    - Split the `odometer_km` into groups, and use aggregation to see if average prices follow any patterns based on the mileage
    - Explore if cars with damage are cheaper than their non-damaged counterparts, and if so, by how much